x-pack/plugins/security_solution/server/endpoint/routes/actions/isolation.ts

```diff
@@ -5,81 +5,146 @@
-      });
-    }
+    isolationRequestHandler(endpointContext, true)
   );
 
   // perform UN-isolate
   router.post(
     {
-      path: `/api/endpoint/unisolate`,
-      validate: {
-        body: schema.object({
-          agent_ids: schema.nullable(schema.arrayOf(schema.string())),
-          endpoint_ids: schema.nullable(schema.arrayOf(schema.string())),
-          alert_ids: schema.nullable(schema.arrayOf(schema.string())),
-          case_ids: schema.nullable(schema.arrayOf(schema.string())),
-          comment: schema.nullable(schema.string()),
-        }),
-      },
+      path: UNISOLATE_HOST_ROUTE,
+      validate: HostIsolationRequestSchema,
       options: { authRequired: true },
     },
-    async (context, req, res) => {
-      if (
-        (req.body.agent_ids === null || req.body.agent_ids.length === 0) &&
-        (req.body.endpoint_ids === null || req.body.endpoint_ids.length === 0)
-      ) {
-        return res.badRequest({
-          body: {
-            message: 'At least one agent ID or endpoint ID is required',
-          },
-        });
-      }
-      return res.ok({
+    isolationRequestHandler(endpointContext, false)
+  );
+}
+
+export const isolationRequestHandler = function (
+  endpointContext: EndpointAppContext,
+  isolate: boolean
+): RequestHandler<
+  unknown,
+  unknown,
+  TypeOf<typeof HostIsolationRequestSchema.body>,
+  SecuritySolutionRequestHandlerContext
+> {
+  return async (context, req, res) => {
+    if (
+      (!req.body.agent_ids || req.body.agent_ids.length === 0) &&
+      (!req.body.endpoint_ids || req.body.endpoint_ids.length === 0)
+    ) {
+      return res.badRequest({
         body: {
-          action: '53ba1dd1-58a7-407e-b2a9-6843d9980068',
+          message: 'At least one agent ID or endpoint ID is required',
         },
       });
     }
-  );
-}
+
+    // only allow admin users
+    const user = endpointContext.service.security?.authc.getCurrentUser(req);
+    if (!UserCanIsolate(user?.roles)) {
+      return res.forbidden({
+        body: {
+          message: 'You do not have permission to perform this action',
+        },
+      });
+    }
+
+    // isolation requires plat+
+    if (isolate && !endpointContext.service.getLicenseService()?.isPlatinumPlus()) {
+      return res.forbidden({
+        body: {
+          message: 'Your license level does not allow for this action',
+        },
+      });
+    }
+
+    // translate any endpoint_ids into agent_ids
+    let agentIDs = req.body.agent_ids?.slice() || [];
+    if (req.body.endpoint_ids && req.body.endpoint_ids.length > 0) {
+      const newIDs = await GetAgentIDsForEndpoints(req.body.endpoint_ids, context, endpointContext);
+      agentIDs = agentIDs.concat(newIDs);
+    }
+
+    // create an Action ID and dispatch it to ES & Fleet Server
+    const esClient = context.core.elasticsearch.client.asCurrentUser;
+    const actionID = uuid.v4();
+    const result = await esClient.index({
+      index: '.fleet-actions',
```

Comment:
I think Fleet exposes this const from their side - I would use that instead of hardcoding the index name